File: tests/Generator/resource/typescript/complex/DefinitionType.ts
import type {ArrayDefinitionType} from "./ArrayDefinitionType";
import type {MapDefinitionType} from "./MapDefinitionType";
import type {StructDefinitionType} from "./StructDefinitionType";

/**
 * Base definition type
 */
export interface DefinitionType {
    deprecated?: boolean
    description?: string
    type?: string
}

File: tests/Generator/resource/csharp/complex/DefinitionType.cs
using System.Text.Json.Serialization;

namespace TypeAPI.Model;

/// <summary>
/// Base definition type
/// </summary>
[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[JsonDerivedType(typeof(ArrayDefinitionType), typeDiscriminator: "array")]
[JsonDerivedType(typeof(MapDefinitionType), typeDiscriminator: "map")]
[JsonDerivedType(typeof(StructDefinitionType), typeDiscriminator: "struct")]
public abstract class DefinitionType
{
    [JsonPropertyName("deprecated")]
    public bool? Deprecated { get; set; }

    [JsonPropertyName("description")]
    public string? Description { get; set; }

    [JsonPropertyName("type")]
    public string? Type { get; set; }

}

